src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -413,7 +428,8 @@ private int FindEntry(TKey key)
                 }
             }
 
-            return i;
+            isFound = false;
+            return ref Unsafe.AsRef<TValue>(null);
```

Comment:
@jkotas I know you don't like `Unsafe` in Collections; would this be a step too far?

Reply:
Could do `private bool FindEntry(TKey key, ref TValue value)` instead?

Reply:
Could do `private bool FindEntry(TKey key, ref TValue value)` instead? 

However, that would mean an unnecessary struct copy for `ContainsKey`

Reply:
FWIW, one could imagine having two additional helper methods for dealing with references:

```cs
static class TheseAreActuallyFullyTypeSafe
{
    bool IsNullRef<T>(in T);
    ref T CreateNullRef<T>();
}
```

If these methods existed this PR wouldn't introduce any "unsafe" code at all into the collection types.

Reply:
Added https://github.com/dotnet/coreclr/pull/27149/commits/0ada3a13988cb0b5542134837d056f59741c75c6

Reply:
> I know you don't like Unsafe in Collections; would this be a step too far?

I can be convinced with numbers. Does it make any measurable improvement in throughput?

Reply:
This doesn't work in shared generics 😢  (including the Span changes)
```
call     [CORINFO_HELP_READYTORUN_GENERIC_HANDLE]
mov      rcx, rax
call     [Unsafe:NullRef():byref]
```

Reply:
NullRef and IsNullRef would have to be leaf methods with IL patched by the VM for this to work.

Reply:
Checking `ResourceManager:InternalGetResourceSet(ref,bool,bool):ref:this` it regresses because `Dictionary`2:TryGetValue(ref,byref):bool:this` now inlines